sweetest/sweetest-api/src/main/java/com/mysugr/sweetest/framework/build/BaseBuilder.kt

```diff
@@ -0,0 +1,266 @@
+package com.mysugr.sweetest.framework.build
+
+import com.mysugr.sweetest.usecases.configureDependencyProvisionAutomatic
+import com.mysugr.sweetest.usecases.configureDependencyMock
+import com.mysugr.sweetest.usecases.configureDependencyProvision
+import com.mysugr.sweetest.usecases.configureDependencyReal
+import com.mysugr.sweetest.usecases.configureDependencySpy
+import com.mysugr.sweetest.usecases.notifyStepsRequired
+import com.mysugr.sweetest.framework.base.SweetestException
+import com.mysugr.sweetest.framework.configuration.ModuleTestingConfiguration
+import com.mysugr.sweetest.framework.context.TestContext
+import com.mysugr.sweetest.framework.dependency.DependencyProvider
+import com.mysugr.sweetest.framework.dependency.asCoreDependencyProvider
+import com.mysugr.sweetest.framework.flow.InitializationStep
+import com.mysugr.sweetest.internal.Steps
+import com.mysugr.sweetest.usecases.subscribeWorkflow
+import kotlin.reflect.KClass
+
+private const val DEPENDENCY_MODE_DEPRECATION_MESSAGE = "Dependency modes like \"REAL\" or \"MOCK\" " +
+    "as well as \"required...\" are obsolete. Use \"provide\" instead."
+
+abstract class BaseBuilder<TSelf>(
+    internal val testContext: TestContext,
+    internal val moduleTestingConfiguration: ModuleTestingConfiguration?
+) {
+
+    private var isFrozen = false
+
+    internal fun freeze() {
+        checkNotYetFrozen()
+        isFrozen = true
+    }
+
+    protected fun checkNotYetFrozen() {
+        if (isFrozen) {
+            error("Can't apply configuration after configuration is finished")
+        }
+    }
```

Comment:
i don't know whether this can really be called a sub-lifecycle. `freeze` is called after `configure()` has been called. after that it can't be called anymore without causing an exception. but to be really honest i wanted to concentrate the efforts to delivering v2 instead of backporting the way of configuration (if you mean that by "unifying"). in which way do you think this can become complicated?

Reply:
> in which way do you think this can become complicated?

The lifecycle of `BaseBuilder` looks like this:
```
┌──────────┐          ┌──────────┐
│ unforzen │─────────>│  frozen  │
└──────────┘  freeze  └──────────┘
```
As I understood it, the idea is to prevent that the configuration cannot be changed anymore just before the tests start. That means configuration should be frozen when `WorkflowSteps` moves to `INITIALIZE_STEPS`. My thought was now, instead of introducing a new concept of `unfrozen` and `frozen`, `BaseBuilder` could leverage the already existing concept of `WorkflowSteps`.

It was just a thought, no need to change it. Not sure if I got it right and also not sure if it's a good idea. I'm fine either way.

Reply:
thanks for outlining so precisely what you meant! i think there might be a misunderstanding. the freeze is here so you don't call the `BaseBuilder` another time. that can happen _independently_ of the overarching workflow (although it's very unlikely).